src/Library/Lod/LodReader.cpp

```diff
@@ -0,0 +1,228 @@
+        dir_read_ptr += read_size;
+
+        LodDirectory dir;
+        dir.name = (char*)img.filename;
+        dir.file_headers_offset = img.data_offset;
+        _lod_read_directory_files(fp, version, dir, img.num_items);
+
+        dirs.push_back(dir);
+    }
+
+    out_index = dirs;
+    return true;
+}
+
+
+std::unique_ptr<LodReader> LodReader::open(const std::string &filename) {
+    auto lod = std::make_unique<LodReader>();
+    if (nullptr == lod) {
+        return nullptr;
+    }
+
+    auto fp = fopen(filename.c_str(), "rb");
+    if (nullptr == fp) {
+        return nullptr;
+    }
+
+    size_t num_expected_directories = 0;
+    bool is_lod = _lod_parse_header(fp, lod->_version, lod->_description, num_expected_directories);
+    if (!is_lod) {
+        fclose(fp);
+        return nullptr;
+    }
+
+    bool is_index_ok = _lod_read_index(fp, lod->_version, num_expected_directories, lod->_index);
+
+    if (is_index_ok) {
+        lod->_fp = fp;
+        return lod;
+    }
+
+    fclose(fp);
+    return nullptr;
+}
+
+
+bool LodReader::exists(const std::string &filename) const {
+    const auto &dir = _index.front(); // only first dir is ever used, no matter names
+    return dir.files.cend() != std::find_if(
+        dir.files.cbegin(),
+        dir.files.cend(),
+        [&](const LodFile &file) {return iequals(file.name, filename); }
+    );
+}
+
+
+Blob LodReader::read(const std::string &filename) {
+    const auto &dir = _index.front(); // only first dir is ever used, no matter names
+    const auto &file = std::find_if(
+        dir.files.cbegin(),
+        dir.files.cend(),
+        [&](const LodFile &file) {return iequals(file.name, filename); }
+    );
+
+    if (dir.files.cend() == file) {
+        return Blob();
+    }
+
+    fseek(_fp, file->data_offset, SEEK_SET);
+    if (_is_file_compressed(*file)) {
+        LodFileCompressionHeader_Mm6 header;
+        assert(1 == fread(&header, sizeof(header), 1, _fp));
+
+        if (header.decompressed_size) {
+            Blob result = Blob::Allocate(header.compressed_size);
+            fread(result.data(), 1, header.compressed_size, _fp);
+            return zlib::Uncompress(result, header.decompressed_size);
+        } else {
+            Blob res = Blob::Allocate(header.compressed_size);
+            assert(1 == fread(res.data(), res.size(), 1, _fp));
+        }
+    }
+    
+    Blob res = Blob::Allocate(file->size);
+    assert(1 == fread(res.data(), res.size(), 1, _fp));
+
+    return res;
+}
+
+
+bool LodReader::_is_file_compressed(const LodFile& file) {
+    if (file.size <= sizeof(LodFileCompressionHeader_Mm6)) {
+        return false;
+    }
+
+    size_t prev_pos = ftell(_fp);
+    fseek(_fp, file.data_offset, SEEK_SET);
+
+    LodFileCompressionHeader_Mm6 header;
+    assert(1 == fread(&header, sizeof(header), 1, _fp));
+    fseek(_fp, prev_pos, SEEK_SET);
+
+    return header.version == 91969 && !memcmp(header.signature, "mvii", 4);
+}
```

Comment:
the reading logic is done, really, and we can replace old lod files pretty quickly
the only missing part is textures/sprites, but that's just compressed reading + some extra texture specific logic that should have never been in LOD anyways